src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -287,29 +298,40 @@ SafeToPushdownUnionSubquery(PlannerRestrictionContext *plannerRestrictionContext
 										   &partitionKeyIndex);
 
 			/* union does not have partition key in the target list */
-			if (partitionKeyIndex == 0)
+			if (partitionKeyIndex == 0 || varToBeAdded == NULL)
 			{
 				continue;
 			}
 		}
 		else
 		{
-			partitionKeyIndex =
-				RelationRestrictionPartitionKeyIndex(relationRestriction);
+			int targetRTEIndex = GetRTEIdentity(relationRestriction->rte);
```

Comment:
The next obvious question is that why do we have the distinction between certain UNION ALL queries and the rest of UNION/UNION ALL queries. 

Can't we get rid of the above part as well? https://github.com/citusdata/citus/compare/fix/union_pushdown...onder_test?expand=1


I think we should do and try to find a simple correlated UNION ALL test which used to work with the above code, and now works with the new code.

What do you think?

Reply:
Good point.

Some example query that would fail without the above block you shared prior to the fix but passes with the new logic is:

```sql
  SELECT
   (SELECT count(*) FROM users_table_part WHERE user_id = e.user_id
       UNION ALL
    SELECT count(*) FROM users_table_part WHERE user_id = e.user_id)
 FROM
   (SELECT * FROM users_table_part UNION ALL SELECT * FROM users_table_part) as e;
```